lib/web/command.go

```diff
@@ -395,35 +396,21 @@ func outputByName(hosts []hostInfo, output map[string][]byte) map[string][]byte
 }
 
 // runCommands runs the given command on the given hosts.
-func runCommands(hosts []hostInfo, runCmd func(host *hostInfo) error, log logrus.FieldLogger) {
-	// Create a synchronization channel to limit the number of concurrent commands.
-	// The maximum number of concurrent commands is 30 - it is arbitrary.
-	syncChan := make(chan struct{}, 30)
-	// WaiteGroup to wait for all commands to finish.
-	wg := sync.WaitGroup{}
+func runCommands(hosts []hostInfo, runCmd func(host *hostInfo) error, numParallel int64, log logrus.FieldLogger) {
+	var group errgroup.Group
+	group.SetLimit(int(numParallel))
 
 	for _, host := range hosts {
 		host := host
-		wg.Add(1)
-
-		go func() {
-			defer wg.Done()
-
-			// Limit the number of concurrent commands.
-			syncChan <- struct{}{}
-			defer func() {
-				// Release the command slot.
-				<-syncChan
-			}()
-
-			if err := runCmd(&host); err != nil {
-				log.WithError(err).Warnf("Failed to start session: %v", host.hostName)
-			}
-		}()
+		group.Go(func() error {
+			return trace.Wrap(runCmd(&host), "failed to start session on %v", host.hostName)
+		})
 	}
 
 	// Wait for all commands to finish.
-	wg.Wait()
+	if err := group.Wait(); err != nil {
+		log.Warn(err)
```

Comment:
This should be debug or trace. I don't think that having server logs about a user calling incorrect command is useful.